components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt

```diff
@@ -654,22 +654,42 @@ class GeckoEngine(
         }
     }
 
+    private fun isFeatureActive(category: TrackingCategory) =
+            (settings.trackingProtectionPolicy?.contains(category)
+                    ?: false)
     /**
      * Mimics the behavior for categorizing trackers from desktop, they should be kept in sync,
      * as differences will result in improper categorization for trackers.
      * https://dxr.mozilla.org/mozilla-central/source/browser/base/content/browser-siteProtections.js
      */
     internal fun ContentBlockingController.LogEntry.toTrackerLog(): TrackerLog {
         val cookiesHasBeenBlocked = this.blockingData.any { it.hasBlockedCookies() }
-        return TrackerLog(
-            url = origin,
-            loadedCategories = blockingData.map { it.getLoadedCategory() }
-                .filterNot { it == TrackingCategory.NONE }
-                .distinct(),
-            blockedCategories = blockingData.map { it.getBlockedCategory() }
+        val blockedCategories = blockingData.map { it.getBlockedCategory() }
                 .filterNot { it == TrackingCategory.NONE }
-                .distinct(),
-            cookiesHasBeenBlocked = cookiesHasBeenBlocked
+                .distinct()
+        val isShimmed = blockingData.any { it.category == Event.REPLACED_TRACKING_CONTENT }
+
+        // When a resource is shimmed, we get both events a REPLACED_UNSAFE_CONTENT
+        // and a loaded event, as the content is blocked but replaced with a shimmed file,
+        // we need to move the event from the allowed list to the blocked list to
+        // reflect the shimmed status.
+        val shimmedCategories = mutableListOf<TrackingCategory>()
+        fun isAddedToShimmedCategories(category: TrackingCategory): Boolean {
+            return if (isFeatureActive(category) && isShimmed) {
+                shimmedCategories.add(category)
+            } else false
+        }
+
+        val loadedCategories = blockingData.map { it.getLoadedCategory() }
+                .filterNot {
+                    it == TrackingCategory.NONE || isAddedToShimmedCategories(it)
+                }.distinct()
+
+        return TrackerLog(
+                url = origin,
+                loadedCategories = loadedCategories,
+                blockedCategories = (blockedCategories + shimmedCategories).distinct(),
+                cookiesHasBeenBlocked = cookiesHasBeenBlocked
```

Comment:
I think we can write this whole block in a more concise way using immutable collections only:

```kotlin
val loadedCategories = blockingData
  .map {
    it.getLoadedCategory()
  }
  .distinct()

val shimmedCategories = loadedCategories
  .filter {
    it == TrackingCategory.NONE || (isFeatureActive(it) && isShimmed)
  }
  .distinct()

return TrackerLog(
    url = origin,
    loadedCategories = loadedCategories - shimmedCategories,
    blockedCategories = (blockedCategories + shimmedCategories).distinct(),
    cookiesHasBeenBlocked = cookiesHasBeenBlocked
)
```

Tests pass fine with this. Wdyt?